### DISCUSSION TITLE
After updating the Bluetooth device, fwupd shows "failed to open /dev/hidraw: No such file or directory"

### DISCUSSION THREAD

**Discussion by sam412081go** [fwupd/fwupd#7343] [Q&A] [answered]
@hughsie 
I used a Chromebook for firmware update testing, but the result failed with the message `failed to open device for cleanup: failed to open /dev/hidraw3: No such file or directory`. 
However, when I used fwupdmgr get-devices to check my BT mouse, I could see that the version number had successfully upgraded. 
[chrome_fwupdmgr_update.txt](https://github.com/user-attachments/files/15836095/chrome_fwupdmgr_update.txt)
[chrome_fwupdtool_update.txt](https://github.com/user-attachments/files/15836096/chrome_fwupdtool_update.txt) => for detail
The fwupd version and ChromeOS version is:
[chrome_fwupd_version.txt](https://github.com/user-attachments/files/15836093/chrome_fwupd_version.txt)

I tested this on three different Chromebooks(different OS version、different fwupd version、different brands of laptop), and all of them produced the same result.

However, this issue does not occur on the other one Ubuntu computer.
[ubuntu_fwupdmgr_update.txt](https://github.com/user-attachments/files/15836097/ubuntu_fwupdmgr_update.txt)
[ubuntu_fwupdtool_update.txt](https://github.com/user-attachments/files/15836092/ubuntu_fwupdtool_update.txt) => for detail

Since the final step of the update process sends a reset command to the mouse to restart it, this action causes the mouse to reconnect with the computer via Bluetooth. At this point, the /dev/hidraw* device disappears. I suspect this is the reason why fwupd cannot find the device.

**Comment by hughsie** [comment, marked as answer]
If the device is re-enumerating I think you need to do something like:

	diff --git a/plugins/pixart-rf/fu-pxi-ble-device.c b/plugins/pixart-rf/fu-pxi-ble-device.c
	index a8b604ef3..aaad1ac46 100644
	--- a/plugins/pixart-rf/fu-pxi-ble-device.c
	+++ b/plugins/pixart-rf/fu-pxi-ble-device.c
	@@ -803,6 +803,7 @@ fu_pxi_ble_device_write_firmware(FuDevice *device,
		if (!fu_pxi_ble_device_reset(self, error))
			return FALSE;
		fu_progress_step_done(progress);
	+       fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
	 
		/* success */
		return TRUE;

I have no idea how it works on Ubuntu without that flag set TBH. I guess Ubuntu is slightly faster and does the cleanup before the device actually reboots.

**Reply by sam412081go** [reply]
Ah~ I see, I will try this. Thanks for your help!

**Reply by sam412081go** [reply]
@hughsie 
I added `fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);` after reset command as below:
![Screenshot from 2024-06-17 11-24-51](https://github.com/fwupd/fwupd/assets/17995855/e7c49241-e74c-4a2a-9068-ccac950bf731)

It will cause update fail, shows `failed to get device before update attach: device ID dffc8f10a1b8128add07b7cd2b610bf11540ad5f was not found`

[log_20240617.txt](https://github.com/user-attachments/files/15861063/log_20240617.txt)

I test it on My Ubuntu computer.

**Reply by sam412081go** [reply]
@hughsie 
Do you have any idea about this?

**Comment by sam412081go** [comment]
fixed in https://github.com/fwupd/fwupd/pull/7361
